### PR TITLE
feat(oauth): make server compatible with AppAuth

### PR DESCRIPTION
**Pull request by vladikoff**

### CODE REVIEW

**Comment by GitCop** [comment]
Thanks for contributing! Unfortunately, I'm here to tell you there were the following style issues with your Pull Request:

 * Commit: 5964ac8f78ce13895467b5218eb78d725ffdaf7a
  * Commits must be in the following format: **%{type}(%{scope}): %{description}**

Guidelines are available at https://github.com/mozilla/fxa-oauth-server

---
This message was auto-generated by https://gitcop.com

**Comment by GitCop** [comment]
Thanks for contributing! Unfortunately, I'm here to tell you there were the following style issues with your Pull Request:

 * Commit: 5964ac8f78ce13895467b5218eb78d725ffdaf7a
  * Commits must be in the following format: **%{type}(%{scope}): %{description}**

 * Commit: de191340b4d3ad67156c580303809480058c0eac
  * Commits must be in the following format: **%{type}(%{scope}): %{description}**

Guidelines are available at https://github.com/mozilla/fxa-oauth-server

---
This message was auto-generated by https://gitcop.com

File: lib/routes/token.js [hunk 1]
```diff
@@ -83,7 +83,7 @@ const PAYLOAD_SCHEMA = Joi.object({
       is: GRANT_JWT,
       then: Joi.forbidden()
     }),
-
+  redirect_uri: Joi.string().optional(),
```
Comment (rfk) [inline lib/routes/token.js]: I suggest we do a little more validation here, e.g. checking that it doesn't contain any weirdy control characters etc.

**Comment by GitCop** [comment]
Thanks for contributing! Unfortunately, I'm here to tell you there were the following style issues with your Pull Request:

 * Commit: db7b5d4e51b80717c8650b580a44b2f05269d264
  * Commits must be in the following format: **%{type}(%{scope}): %{description}**

 * Commit: e377697a07a75fe04a134fa1b03e5f7dd4d9bfa2
  * Commits must be in the following format: **%{type}(%{scope}): %{description}**

Guidelines are available at https://github.com/mozilla/fxa-oauth-server

---
This message was auto-generated by https://gitcop.com

**Comment by vladikoff** [comment]
@rfk r? I tested the refresh token flows with AppAuth and those worked great. We just need to land these changes to make it compatible. Let me know!

File: lib/validators.js [hunk 2]
```diff
@@ -32,6 +32,10 @@ exports.scope = Joi.string()
   .max(256)
   .regex(/^[a-zA-Z0-9 _\/.:]+$/);
 
+exports.redirectUri = Joi.string()
+  .max(256)
+  .regex(/^[a-zA-Z0-9\-_\/.:]+$/);
```
Comment (rfk) [inline lib/validators.js]: The auth-server has a URI validator that we could copy for this, but I'm happy for that to be a follow-up.

Comment (vladikoff) [reply lib/validators.js]: Did you mean https://github.com/mozilla/fxa-auth-server/blob/10e934f5fe71c464ffb0bcf1b71f5b61fd4afc47/lib/routes/validators.js#L176 ?

Comment (rfk) [reply lib/validators.js]: Yes, and https://github.com/mozilla/fxa-auth-server/blob/10e934f5fe71c464ffb0bcf1b71f5b61fd4afc47/lib/routes/validators.js#L158 that uses it

Comment (vladikoff) [reply lib/validators.js]: Hm at first glance it seems using that will break something like `testpilot-thing://oauth-redirect`

Comment (rfk) [reply lib/validators.js]: Hrm, indeed; *shrug* I don't fell that strongly about it TBH so happy to just leave it for now :-)

Comment (l-hedgehog) [reply lib/validators.js]: This pattern prohibits query strings in redirectUri, we have a few oauth relier with such redirectUris in China, is it an intentional limit or not?

Comment (l-hedgehog) [reply lib/validators.js]: Filed as #547, thanks!

**Pull request merged**